### PR TITLE
[GSoC 2022] spng encoder/decoder added as optional png codec

### PR DESCRIPTION
**Pull request by ocpalo**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/995

This pull request contains the implementation of PngEncoder using spng library. One thing that I am not sure about is the CMake configuration. I appreciate any help on CMake configuration

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (vpisarev) [inline modules/imgcodecs/src/grfmt_spng.cpp:253]: 1. why such a big buffer for the whole image is allocated?
2. it's not deallocated, it's a memory leak. Please, do not allocate memory manually unless it's necessary. Use std::vector.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (vpisarev) [inline modules/imgcodecs/src/grfmt_spng.cpp:246]: is it necessary to call this function? You decode the image row by row after it in any case.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (vpisarev) [inline modules/imgcodecs/src/grfmt_spng.cpp:243]: the code is correct, but still quite difficult to read. Could you please use explicit `if (ret == SPNG_OK) { ... }` to make the code easier to read?

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (vpisarev) [inline modules/imgcodecs/src/grfmt_spng.cpp:134]: where is ctx released?

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (vpisarev) [inline modules/imgcodecs/src/grfmt_spng.cpp:316]: the API is unstable. You call close() after data is read. User may want, for some reason, call readData several times.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp:347]: If if understand it correctly, there are related tests in OpenCV for that. It encodes to the buffer and compares with some Vector.  All the test cases is passed. 

The answer to your question is in line 407, it checks if the encoder should write it to buffer or file. I copied that part from the libpng codec.

```cpp
            if( m_buf )
            {
                spng_set_png_stream(ctx,(spng_rw_fn*)writeDataToBuf, this);
            }
            else
            {
                f = fopen( m_filename.c_str(), "wb" );
                if( f )
                    spng_set_png_file( ctx, f );
            }
```

Here spng_set_png_stream function call does that.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp:134]: ctx is assigned to the m_ctx. Which is a member variable of the SPngDecoder class. It deallocates it in close function after readHeader function.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp]: This part of the code is left when I am trying to do the necessary conversion of color formats. That was a bad mistake :). Sorry. Fixed.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp]: Done

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp]: I changed it to
```cpp
        if(!result)
            close();
```

Should I create a quick patch for the current PNG decoder too?

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (asmorkalov) [inline modules/imgcodecs/src/grfmt_spng.cpp:137]: Please check if `spng_ctx_free` with NULL is expected behaviour.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (asmorkalov) [inline modules/imgcodecs/src/grfmt_spng.cpp:145]: Why?

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp:137]:
```cpp
void spng_ctx_free(spng_ctx *ctx)
{
    if(ctx == NULL) return;
```

It is expected behavior but it seems unnecessary. For the future releases of spng, it might be useful to keep this line.

File: modules/imgcodecs/src/grfmt_spng.cpp
Comment (ocpalo) [reply modules/imgcodecs/src/grfmt_spng.cpp:145]: Author of the libspng wrote this in the example for the setting the chunk limit.

```cpp
    /* Set memory usage limits for storing standard and unknown chunks,
       this is important when reading untrusted files! */
```

I thought it was a good idea. But this limits the chunks size. Now I think it is better if I remove it. This methods equivalent is `png_set_chunk_malloc_max`.

Should i keep this or remove it?

Comment (asmorkalov) [reply modules/imgcodecs/src/grfmt_spng.cpp:145]: I do not have strong opinion here. It's ok to have chunk limit. Please add note why you use the particular value or link to original code, if it's inherited behaviour.

**Comment by ocpalo** [comment]
I did some performance tests and right now it performs worse than the current png codec. I am going to look more at how I could improve it. Let's not merge this one yet.

**Comment by ocpalo** [comment]
Iam thinking to rewrite code as c-style. Let me know if opencv okay with that style.

Example:

```cpp
     int trns = spng_get_trns(...);
     if(trns != SPNG_OK) 
     {
         goto error;
     }

     ...
error:
   //handle error here

   return success;
```

getting rid of nested if statements might improve performance.

**Comment by ocpalo** [comment]
SPNG perf:
[ PERFSTAT ]    (samples=13   mean=11.52   median=11.42   min=11.25   stddev=0.29 (2.5%))

image_2022-07-27_011002323

PNG perf:
[ PERFSTAT ]    (samples=10   mean=13.22   median=13.19   min=12.99   stddev=0.18 (1.4%))

Screenshot from 2022-07-27 01-12-57

Well interesting. I upgraded my Ubuntu to 22.04 and it seems SPNG slightly performs better now. 

And i changed only one line. `spng_ctx_new( SPNG_CTX_IGNORE_ADLER32 );` I ignore ADLER32 checksum in deflate streams.

**Comment by ocpalo** [comment]
Potential performance improvement: Convert RGB to BGR while decoding image row-by-row. All test cases are passed but ROI still might not work.

```cpp
 ret = spng_decode_row(png_ptr, img.data + row_info.row_num * image_width, image_width);
 ```
 
 Add img.step here.

**Comment by ocpalo** [comment]
@asmorkalov can you run the workflow? Lets see if spng works on other platforms.

**Comment by ocpalo** [comment]
cmake release mode png
[ RUN      ] PNG.decode
[ PERFSTAT ]    (samples=10   mean=179.53   median=178.60   min=177.07   stddev=2.56 (1.4%))

cmake release mode spng
[ RUN      ] PNG.decode
[ PERFSTAT ]    (samples=10   mean=151.74   median=150.96   min=150.39   stddev=1.60 (1.1%))

**Comment by vpisarev** [comment]
@ocpalo, I tested it on my machine; works well well 👍

**Review by vpisarev — APPROVED** [review]

**Comment by ocpalo** [comment]
Before merge, let me revert cmake changes and make spng optional decoder.

**Comment by ocpalo** [comment]
Ubuntu2004-x64 - macOS-ARM/x64 Error:
```cpp
[ RUN      ] Test_ONNX_layers.ReduceSum/0, where GetParam() = OCV/CPU
unknown file: Failure
C++ exception with description "OpenCV(4.6.0-dev) /home/ci/opencv/modules/ts/src/ts.cpp:1064: error: (-2:Unspecified error) OpenCV tests: Can't find required data file: dnn/onnx/models/reduce_sum_axis_dynamic_batch.onnx in function 'findData'
" thrown in the test body.
```

Windows Error:
Fails to build spng library. 

I need to figure out how to compile libspng in Windows environment

**Comment by sturkmen72** [comment]
reminder:
i think in build information we should see if linspng used for png codec

currently we see
in linux
`PNG:                         /usr/lib/aarch64-linux-gnu/libpng.so (ver 1.6.37)`

in windows
`PNG:                         build (ver 1.6.37)`

**Comment by ocpalo** [comment]
Current issues : 
* Gray <= 8 Gamma correction is not implemented by SPNG. Gamma tests of pngsuite fails. 
* Some IMREAD_GRAYSCALE images fails by 1 value. Need to revisit spngCvt functions

**Comment by ocpalo** [comment]
Windows fix: https://ci.appveyor.com/project/sturkmen72/opencv-image-reading/builds/44465042

**Comment by ocpalo** [comment]
Perf test results:
* SPNG
```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from PNG
[ RUN      ] PNG.decode
[ PERFSTAT ]    (samples=10   mean=145.95   median=144.70   min=144.31   stddev=2.76 (1.9%))
[       OK ] PNG.decode (1460 ms)
[ RUN      ] PNG.encode
[ PERFSTAT ]    (samples=10   mean=413.33   median=415.09   min=386.22   stddev=10.20 (2.5%))
[       OK ] PNG.encode (4289 ms)
[----------] 2 tests from PNG (5749 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (5749 ms total)
[  PASSED  ] 2 tests.
```
* PNG
```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from PNG
[ RUN      ] PNG.decode
[ PERFSTAT ]    (samples=10   mean=172.86   median=172.21   min=169.97   stddev=3.23 (1.9%))
[       OK ] PNG.decode (1729 ms)
[ RUN      ] PNG.encode
[ PERFSTAT ]    (samples=25   mean=361.06   median=360.23   min=344.85   stddev=5.38 (1.5%))
[       OK ] PNG.encode (9402 ms)
[----------] 2 tests from PNG (11131 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (11131 ms total)
[  PASSED  ] 2 tests.
```

**Comment by ocpalo** [comment]
@vpisarev @asmorkalov Can we run the workflow once again?

**Comment by ocpalo** [comment]
Last workflow fails but this does not related with my pull request. I added libspng as optional codec. It is not enabled by default. If you look cmake build, libpng is used as png codec.

File: cmake/OpenCVFindLibsGrfmt.cmake
Comment (asmorkalov) [inline cmake/OpenCVFindLibsGrfmt.cmake:267]: I propose to uncomment the line to not have 2 codecs at a time.

**Review by asmorkalov — APPROVED** [review]

File: cmake/OpenCVFindLibsGrfmt.cmake
Comment (ocpalo) [reply cmake/OpenCVFindLibsGrfmt.cmake:267]: This was intended to surpress CMake warning while running the workflow. I made libspng default png codec to test in the workflow. Before merge, i will uncomment that line and set `OCV_OPTION(WITH_SPNG "Include SPNG support" OFF`. So we can add libspng as optional codec, not default as we discussed in one of the GSoC meetings with the Vadim and Suleyman.

**Comment by ocpalo** [comment]
It seems workflow passed. Please let me know if all is right so I can rebase and fix CMake to make spng as optional decoder.

File: modules/imgcodecs/src/grfmt_spng.hpp
Comment (alalek) [inline modules/imgcodecs/src/grfmt_spng.hpp:15]: License header is defined in contribution guidelines.

File: modules/imgcodecs/src/loadsave.cpp
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp:174]: `#elif defined(HAVE_PNG)`

File: modules/imgcodecs/test/test_png.cpp
Comment (alalek) [inline modules/imgcodecs/test/test_png.cpp:200]: It is better to have single file per test case (you don't want to debug issue of 15th image of the test).

**Comment by ocpalo** [comment]
@alalek I made the changes you requested except single file per test case. I am planning to do it in the future. Right now, because of personal issues, i have really no time to investigate in this. And if someone wants to debug it, he can just insert one line to print filename in the test case.

**Comment by ocpalo** [comment]
Let me know if its ok. I will rebase it if you approve

File: modules/imgcodecs/test/test_png.cpp
Comment (vpisarev) [reply modules/imgcodecs/test/test_png.cpp:200]: @alalek, I already investigated this issue. Since it's many tiny files, one big file is probably preferable for now. I suggest to merge as-is

Comment (alalek) [reply modules/imgcodecs/test/test_png.cpp:200]: This is a general of for writing tests. All-in-one test is anti-pattern.

They should be spitted:
- again, as I mentioned - we need possibility to debug exact case without skipping other 10 cases (debugging code is time consuming task, no need to make it harder)
- test code may do nothing if `glob()` returns nothing due to misconfiguration - and we will not see this in test logs
- all-in-one approach blocks test filtering, disabling

File: 3rdparty/libspng/CMakeLists.txt
Comment (alalek) [inline 3rdparty/libspng/CMakeLists.txt:10]: > `,`

Does it work at all? There is no commas for arguments in CMake scripts.

Comment (alalek) [inline 3rdparty/libspng/CMakeLists.txt:10]: > ZLIB_INCLUDE_DIRS

There is dependency on ZLIB, which could be disabled.
Need to properly handle case if zlib is disabled.

File: cmake/OpenCVFindLibsGrfmt.cmake
Comment (alalek) [inline cmake/OpenCVFindLibsGrfmt.cmake:243]: Please minimize amount of change code:

```
if(WITH_SPNG)
   ... new code
endif()
if(NOT HAVE_SPNG AND WITH_PNG)
   ... existed untouched code ...
endif()
```
```

Comment (vpisarev) [reply cmake/OpenCVFindLibsGrfmt.cmake:243]: @alalek, instead of putting generic comments, could you please put the snippet what you have in mind? I personally cannot reverse engineer it from your comment

File: 3rdparty/libspng/CMakeLists.txt
Comment (vpisarev) [reply 3rdparty/libspng/CMakeLists.txt]: fixed

Comment (alalek) [reply 3rdparty/libspng/CMakeLists.txt]: If it doesn't work then probably it is just not needed at all.
So `${CURR_INCLUDE_DIR}` could be dropped completely.

Comment (ocpalo) [reply 3rdparty/libspng/CMakeLists.txt]: Is not this handled in FindLibsGrfmt CMakeList already?

**Comment by ocpalo** [comment]
I rebased all of my commits with the changes that alalek requested on CMakeLists. I am working on to update tests for pngsuite. I will probably finish by the noon, tomorrow.

**Comment by ocpalo** [comment]
Few linux commands did the trick. Here we go. Thanks to @sturkmen72 for separate xml files. He was generated those files while ago and shared with me. @alalek @vpisarev could you review if this is ok now?

File: 3rdparty/libspng/CMakeLists.txt
Comment (alalek) [reply 3rdparty/libspng/CMakeLists.txt]: ZLIB is always "required" in this file.

File: modules/imgcodecs/test/test_png.cpp
Comment (alalek) [inline modules/imgcodecs/test/test_png.cpp:207]: Why do we need this `if` check?

File: modules/imgcodecs/test/test_png.cpp
Comment (alalek) [inline modules/imgcodecs/test/test_png.cpp:386]: > `pngsuite/`

is a common prefix here. It could be moved to the test code

File: modules/imgcodecs/test/test_png.cpp
Comment (alalek) [inline modules/imgcodecs/test/test_png.cpp:389]: > PngSUiteFiles

PngSuiteFiles or `/*nothing*/` (to avoid any prefix with duplicate naming).

File: 3rdparty/libspng/CMakeLists.txt
Comment (ocpalo) [reply 3rdparty/libspng/CMakeLists.txt]: Can you please provide more information about how to handle it? I am not that familiar with CMake and dependency handling. @alalek

**Comment by ocpalo** [comment]
@alalek I have updated the test_png files. But as i said above, i did not understand how to handle zlib dependency handling if ZLIB is disabled. If you provide me more information, i am happy to do it :)

**Review by alalek — APPROVED** [review]
Thank you for contribution!

**Comment by ocpalo** [comment]
fyi, i force pushed to fix docs warnings(whitespaces)

**Comment by ocpalo** [comment]
It seems the runflow fails on some flows. However, when i check the build logs, it builds with the libpng since i have made it libspng as default. I think the workflow fails are not related with my pullrequest.

**Comment by ocpalo** [comment]
You can check this [workflow](https://github.com/opencv/opencv/actions/runs/2885124970) where i enabled libspng. Since then, i have not made any changes in the code except alalek cmake and test requests. @asmorkalov .

**Comment by vpisarev** [comment]
@asmorkalov, looks like the failures have no relation with the patch. Can we finally merge it?

**Comment by asmorkalov** [comment]
I tested the PR manually with ARMv7 with Ubuntu 14.04 and x86_64 with Ubuntu 18.04.
- Pushed new commit to enable EXIF rotation test. It's supported by libspng correctly.
- I continuously get 2 test failures with my Ubuntu 18.04. 2 images from corrupted list are decoded 
```
[ RUN      ] Imgcodecs_Png_PngSuite_Corrupted.decode/3, where GetParam() = "xcsn0g01"
/mnt/projects/Projects/OpenCV/opencv-master/modules/imgcodecs/test/test_png.cpp:385: Failure
Value of: src.empty()
  Actual: false
Expected: true
[  FAILED  ] Imgcodecs_Png_PngSuite_Corrupted.decode/3, where GetParam() = "xcsn0g01" (0 ms)
[ RUN      ] Imgcodecs_Png_PngSuite_Corrupted.decode/4, where GetParam() = "xd0n2c08"
[       OK ] Imgcodecs_Png_PngSuite_Corrupted.decode/4 (0 ms)
[ RUN      ] Imgcodecs_Png_PngSuite_Corrupted.decode/5, where GetParam() = "xd3n2c08"
[       OK ] Imgcodecs_Png_PngSuite_Corrupted.decode/5 (0 ms)
[ RUN      ] Imgcodecs_Png_PngSuite_Corrupted.decode/6, where GetParam() = "xd9n2c08"
[       OK ] Imgcodecs_Png_PngSuite_Corrupted.decode/6 (0 ms)
[ RUN      ] Imgcodecs_Png_PngSuite_Corrupted.decode/7, where GetParam() = "xdtn0g01"
[       OK ] Imgcodecs_Png_PngSuite_Corrupted.decode/7 (0 ms)
[ RUN      ] Imgcodecs_Png_PngSuite_Corrupted.decode/8, where GetParam() = "xhdn0g08"
/mnt/projects/Projects/OpenCV/opencv-master/modules/imgcodecs/test/test_png.cpp:385: Failure
Value of: src.empty()
  Actual: false
Expected: true
[  FAILED  ] Imgcodecs_Png_PngSuite_Corrupted.decode/8, where GetParam() = "xhdn0g08" (0 ms)
```
`xcsn0g01.png` works well with OpenCV (imshow) and my KDE image viewer.
`xhdn0g08.png` shows blury lines with OpenCV and KDE viewer reports corrupted image.
The issue is reproducible with both system zlib and self build (-DBUILD_ZLIB=ON).

**Comment by ocpalo** [comment]
@asmorkalov that is a very good catch. I was not able to catch this because I have not tested after changing unit test structure. I assumed it will pass as before but it seems alalek was right to change to catch this kind of issues. 

About the issue, in the implementation, I ignore the checksum of IHDR and IDAT. I prefer to do it that way because many libraries ignores it in their implementation. Since it does not calculate the checksum, it decodes the images into mat object. How should I approach this situation? Ignore the calculation of checksum and change unit tests or should I modify the existing code to calculate the checksum? I think it is better you, opencv team, decide how to handle this situation. Downside of calculation checksum is it reduces the performance.

**Comment by asmorkalov** [comment]
It's interesting question. I do not see issue with Ubuntu 14.04 on my jetson tk1. First of all we need deterministic behavior everywhere. I'm looking on Ubuntu 20.04 right now to find correlation.

**Comment by sturkmen72** [comment]
for information see http://www.schaik.com/pngsuite/pngsuite_xxx_png.html

**Comment by asmorkalov** [comment]
The issue is reproducible with Ubuntu 20.04.

**Comment by ocpalo** [comment]
@asmorkalov to keep the existing behaviour of the current png codec, i have changed to calculate to checksum. Sorry to bother you with this. I have tested on Ubuntu 22.04 and tests are passed now. Could you please review/re-run your tests?

As a side note, we can provide some kind of flags to users to set these. For example, in my latest pull request about new imread function with named-parameters, we can add some kind of flags, boolean, enum etc, to let user to control it. I am not expert at codecs but i believe most of the codecs have this kind of checks, flags etc.

**Comment by vpisarev** [comment]
@asmorkalov, can you check this patch once again and see if the problem is solved on your Ubuntu machine?

**Review by asmorkalov — APPROVED** [review]
:+1: Great job! Thanks a lot for the contribution!

**Pull request merged**
